// varlock/dist/chunk-YBSEQARI.js
import { gracefulExit, getItemSummary, joinAndCompact } from './chunk-PPIQGMSC.js';
import { ansis_default, VarlockError, my_dash_default } from './chunk-CBO5NXPJ.js';
import { __name } from './chunk-6PEHRAEP.js';

// src/cli/helpers/error-checks.ts
function showErrorLocationDetails(err) {
  if (!(err instanceof VarlockError) || !err.location) return;
  const errLoc = err.location;
  const errPreview = [
    errLoc.lineStr,
    `${ansis_default.gray("-".repeat(errLoc.colNumber - 1))}${ansis_default.red("^")}`
  ].join("\n");
  console.log("");
  console.log(`\u{1F4C2} ${errLoc.id}:${errLoc.lineNumber}:${errLoc.colNumber}`);
  console.log(errPreview);
}
__name(showErrorLocationDetails, "showErrorLocationDetails");
function checkForNoEnvFiles(envGraph) {
  if (Object.keys(envGraph.configSchema).length === 0) {
    if (!envGraph.rootDataSource) {
      console.error("\u{1F6A8} No .env or .env.schema files found\n");
      console.error("Run `varlock init` to create a .env.schema file, or use `--path` to specify a file or directory.");
    } else {
      console.error("\u{1F6A8} No items defined in your .env or .env.schema files\n");
      console.error("Add items to your .env.schema file to get started.");
    }
    return gracefulExit(1);
  }
}
__name(checkForNoEnvFiles, "checkForNoEnvFiles");
function checkForSchemaErrors(envGraph) {
  for (const source of envGraph.sortedDataSources) {
    if (source.loadingError) {
      console.log(`\u{1F6A8} Error encountered while loading ${source.label}
`);
      console.log(source.loadingError.message);
      showErrorLocationDetails(source.loadingError);
      if (source.loadingError.stack && !(source.loadingError instanceof VarlockError)) {
        console.log(`
${ansis_default.dim("Stack trace:")}`);
        console.log(ansis_default.dim(source.loadingError.stack));
      }
      return gracefulExit(1);
    }
    if (source.schemaErrors.length) {
      console.log(`\u{1F6A8} Error(s) encountered in ${source.label}`);
      for (const schemaErr of source.schemaErrors) {
        console.log(`- ${schemaErr.message}`);
        showErrorLocationDetails(schemaErr);
      }
      return gracefulExit(1);
    }
  }
}
__name(checkForSchemaErrors, "checkForSchemaErrors");
function showPluginWarnings(envGraph) {
  for (const plugin of envGraph.plugins) {
    if (!plugin.warnings.length) continue;
    for (const warning of plugin.warnings) {
      console.log(ansis_default.yellow(`[WARNING] ${warning.message}`));
      if (warning.tip) {
        for (const line of warning.tip.split("\n")) {
          console.log(`  ${line}`);
        }
      }
    }
  }
}
__name(showPluginWarnings, "showPluginWarnings");
var InvalidEnvError = class extends Error {
  static {
    __name(this, "InvalidEnvError");
  }
  constructor() {
    super("Resolved config/env did not pass validation");
  }
  getFormattedOutput() {
    return `
\u{1F4A5} ${ansis_default.red(this.message)} \u{1F4A5}
`;
  }
};
function checkForConfigErrors(envGraph, opts) {
  for (const source of envGraph.sortedDataSources) {
    if (source.resolutionErrors.length) {
      console.log(`\u{1F6A8} Root decorator error(s) in ${source.label}`);
      for (const err of source.resolutionErrors) {
        console.log(`- ${err.message}`);
        if (err instanceof VarlockError && err.tip) {
          for (const line of err.tip.split("\n")) {
            console.log(`  ${line}`);
          }
        }
        showErrorLocationDetails(err);
      }
    }
  }
  const failingItems = envGraph.sortedConfigKeys.map((k) => envGraph.configSchema[k]).filter((item) => item.validationState === "error");
  if (failingItems.length > 0) {
    console.error(`
\u{1F6A8} \u{1F6A8} \u{1F6A8}  ${ansis_default.bold.underline("Configuration is currently invalid ")}  \u{1F6A8} \u{1F6A8} \u{1F6A8}
`);
    console.error("Invalid items:\n");
    my_dash_default.each(failingItems, (item) => {
      console.error(getItemSummary(item));
      console.error();
    });
    if (opts?.showAll) {
      console.error();
      console.error(joinAndCompact([
        "Valid items:",
        ansis_default.italic.gray("(remove `--show-all` flag to hide)")
      ]));
      console.error();
      const validItems = envGraph.sortedConfigKeys.map((k) => envGraph.configSchema[k]).filter((i) => !!i.isValid);
      my_dash_default.each(validItems, (item) => {
        console.error(getItemSummary(item));
      });
    }
    showPluginWarnings(envGraph);
    throw new InvalidEnvError();
  }
}
__name(checkForConfigErrors, "checkForConfigErrors");

export { InvalidEnvError, checkForConfigErrors, checkForNoEnvFiles, checkForSchemaErrors, showPluginWarnings };
