// varlock/dist/chunk-A2HUIC6U.js
import { define } from './chunk-4A54P4EM.js';
import { checkForSchemaErrors, checkForNoEnvFiles, checkForConfigErrors, showPluginWarnings } from './chunk-YBSEQARI.js';
import { loadVarlockEnvGraph } from './chunk-HSRXSFTK.js';
import { getItemSummary } from './chunk-PPIQGMSC.js';
import { __name } from './chunk-6PEHRAEP.js';

// src/cli/commands/load.command.ts
var commandSpec = define({
  name: "load",
  description: "Load env according to schema and resolve values",
  args: {
    format: {
      type: "enum",
      short: "f",
      choices: ["pretty", "json", "env", "shell", "json-full"],
      description: "Format of output",
      default: "pretty"
    },
    compact: {
      type: "boolean",
      description: "Use compact format (for json-full: no indentation, for env/shell: skip undefined values)"
    },
    "show-all": {
      type: "boolean",
      description: "When load is failing, show all items rather than only failing items"
    },
    env: {
      type: "string",
      description: "Set the environment (e.g., production, development, etc) - will be overridden by @currentEnv in the schema if present"
    },
    path: {
      type: "string",
      short: "p",
      description: "Path to a specific .env file or directory to use as the entry point"
    }
  },
  examples: `
Loads and validates environment variables according to your .env files, and prints the results.
Useful for debugging locally, and in CI to print out a summary of env vars.

Examples:
  varlock load                    # Load and validate with pretty output
  varlock load --format json      # Output in JSON format
  eval "$(varlock load --format shell)"  # Load vars into current shell (useful with direnv)
  varlock load --show-all         # Show all items when validation fails
  varlock load --path .env.prod   # Load from a specific .env file
  varlock load --compact          # Use compact format - skips undefined values, no indentation for json-full
  varlock load --env production   # Load for a specific environment (\u26A0\uFE0F ignored if using @currentEnv!)
`.trim()
});
var commandFn = /* @__PURE__ */ __name(async (ctx) => {
  const { format, compact, "show-all": showAll } = ctx.values;
  const envGraph = await loadVarlockEnvGraph({
    currentEnvFallback: ctx.values.env,
    entryFilePath: ctx.values.path
  });
  checkForSchemaErrors(envGraph);
  checkForNoEnvFiles(envGraph);
  if (!envGraph.rootDataSource) throw new Error("expected root data source to be set");
  await envGraph.generateTypesIfNeeded();
  await envGraph.resolveEnvValues();
  checkForConfigErrors(envGraph, { showAll });
  if (format === "pretty") {
    showPluginWarnings(envGraph);
    for (const itemKey of envGraph.sortedConfigKeys) {
      const item = envGraph.configSchema[itemKey];
      console.log(getItemSummary(item));
    }
  } else if (format === "json") {
    console.log(JSON.stringify(envGraph.getResolvedEnvObject(), null, 2));
  } else if (format === "json-full") {
    const indent = compact ? 0 : 2;
    console.log(JSON.stringify(envGraph.getSerializedGraph(), null, indent));
  } else if (format === "env" || format === "shell") {
    const resolvedEnv = envGraph.getResolvedEnvObject();
    const skipUndefined = compact === true;
    const prefix = format === "shell" ? "export " : "";
    for (const key in resolvedEnv) {
      const value = resolvedEnv[key];
      if (value === void 0 && skipUndefined) {
        continue;
      }
      let strValue;
      if (value === void 0) {
        strValue = "";
      } else if (typeof value === "string") {
        strValue = `"${value.replaceAll('"', '\\"').replaceAll("\n", "\\n")}"`;
      } else {
        strValue = JSON.stringify(value);
      }
      console.log(`${prefix}${key}=${strValue}`);
    }
  } else {
    throw new Error(`Unknown format: ${format}`);
  }
}, "commandFn");

export { commandFn, commandSpec };
