// varlock/dist/chunk-PZZHI3DT.js
import { define } from './chunk-4A54P4EM.js';
import { checkForSchemaErrors, checkForNoEnvFiles, checkForConfigErrors } from './chunk-YBSEQARI.js';
import { loadVarlockEnvGraph } from './chunk-HSRXSFTK.js';
import { gracefulExit } from './chunk-PPIQGMSC.js';
import { resetRedactionMap, redactSensitiveConfig } from './chunk-IMB5QAZS.js';
import { __name } from './chunk-6PEHRAEP.js';
import { spawn } from 'child_process';
import 'stream';
import { extname, isAbsolute, delimiter, join } from 'path';
import { accessSync, constants, statSync, existsSync, readFileSync } from 'fs';

function getWindowsExtensions() {
  const pathExt = process.env.PATHEXT || ".EXE;.CMD;.BAT;.COM";
  const exts = pathExt.split(";").filter(Boolean);
  const result = [];
  for (const ext of exts) {
    result.push(ext.toUpperCase());
    result.push(ext.toLowerCase());
  }
  return result;
}
__name(getWindowsExtensions, "getWindowsExtensions");
function isExecutable(filePath) {
  try {
    accessSync(filePath, constants.X_OK);
    const stats = statSync(filePath);
    return stats.isFile();
  } catch {
    return false;
  }
}
__name(isExecutable, "isExecutable");
function isExecutableOnWindows(filePath) {
  if (!existsSync(filePath)) return false;
  try {
    const stats = statSync(filePath);
    if (!stats.isFile()) return false;
  } catch {
    return false;
  }
  const ext = extname(filePath).toLowerCase();
  const extensions = getWindowsExtensions().map((e) => e.toLowerCase());
  return extensions.includes(ext);
}
__name(isExecutableOnWindows, "isExecutableOnWindows");
function readShebang(filePath) {
  try {
    const fd = readFileSync(filePath, { encoding: "utf8", flag: "r" });
    const first150 = fd.slice(0, 150);
    const match = first150.match(/^#!([^\r\n]+)/);
    return match ? match[1].trim() : null;
  } catch {
    return null;
  }
}
__name(readShebang, "readShebang");
function findCommand(command) {
  const isWin = process.platform === "win32";
  const hasPathSep = command.includes("/") || command.includes("\\");
  if (isAbsolute(command) || hasPathSep) {
    return command;
  }
  const pathEnv = process.env.PATH || "";
  const searchPaths = [];
  if (isWin) {
    searchPaths.push(process.cwd());
  }
  const pathParts = pathEnv.split(delimiter);
  for (const part of pathParts) {
    const cleanPart = /^".*"$/.test(part) ? part.slice(1, -1) : part;
    if (cleanPart) {
      searchPaths.push(cleanPart);
    }
  }
  let extensions = [""];
  if (isWin) {
    extensions = getWindowsExtensions();
    if (command.includes(".") && process.env.PATHEXT) {
      extensions.unshift("");
    }
  }
  for (const dir of searchPaths) {
    for (const ext of extensions) {
      const fullPath = join(dir, command + ext);
      if (isWin) {
        if (isExecutableOnWindows(fullPath)) {
          return fullPath;
        }
      } else {
        if (isExecutable(fullPath)) {
          const shebang = readShebang(fullPath);
          if (shebang && shebang.startsWith("/")) {
            return fullPath;
          } else if (!shebang) {
            return fullPath;
          }
        }
      }
    }
  }
  return command;
}
__name(findCommand, "findCommand");
function exec(command, args, options = {}) {
  const resolvedCommand = findCommand(command);
  const needsShell = process.platform === "win32" && /\.(cmd|bat)$/i.test(resolvedCommand);
  let spawnCommand = resolvedCommand;
  let spawnArgs = args;
  const spawnOptions = {
    env: options.env || process.env,
    shell: false
  };
  if (needsShell) {
    spawnArgs = ["/d", "/s", "/c", `"${resolvedCommand}" ${args.map((a) => `"${a}"`).join(" ")}`];
    spawnCommand = process.env.comspec || "cmd.exe";
    spawnOptions.windowsVerbatimArguments = true;
  }
  let stdio;
  if (options.stdio === "inherit") {
    stdio = "inherit";
  } else if (options.stdio === "pipe") {
    stdio = ["pipe", "pipe", "pipe"];
  } else if (options.stdio) {
    stdio = options.stdio;
  } else {
    stdio = [
      options.stdin || "inherit",
      options.stdout || "inherit",
      options.stderr || "inherit"
    ];
  }
  spawnOptions.stdio = stdio;
  const childProcess = spawn(spawnCommand, spawnArgs, spawnOptions);
  const result = {
    stdout: childProcess.stdout || void 0,
    stderr: childProcess.stderr || void 0,
    pid: childProcess.pid,
    kill: /* @__PURE__ */ __name((signal) => childProcess.kill(signal), "kill")
  };
  const promise = new Promise((resolve, reject) => {
    let errorEmitted = false;
    childProcess.on("error", (error) => {
      errorEmitted = true;
      reject(
        Object.assign(error, {
          exitCode: 1,
          ...result
        })
      );
    });
    childProcess.on("exit", (code, signal) => {
      if (process.platform === "win32" && code === 1 && !errorEmitted && !existsSync(resolvedCommand)) {
        const error = new Error(`Command not found: ${command}`);
        error.code = "ENOENT";
        error.exitCode = 1;
        Object.assign(error, result);
        reject(error);
        return;
      }
      const exitCode = code ?? (signal ? 1 : 0);
      const exitResult = {
        exitCode,
        signal: signal || void 0,
        ...result
      };
      if (exitCode !== 0) {
        const error = new Error(`Command failed with exit code ${exitCode}`);
        error.exitCode = exitCode;
        error.signal = signal;
        Object.assign(error, result);
        reject(error);
      } else {
        resolve(exitResult);
      }
    });
  });
  Object.assign(promise, result);
  return promise;
}
__name(exec, "exec");

// src/cli/commands/run.command.ts
var commandSpec = define({
  name: "run",
  description: "Run a command with your environment variables injected",
  args: {
    // watch: {
    //   type: 'boolean',
    //   short: 'w',
    //   description: 'Watch mode',
    // },
    "no-redact-stdout": {
      type: "boolean",
      description: "Disable stdout/stderr redaction to preserve TTY detection for interactive tools"
    },
    path: {
      type: "string",
      short: "p",
      description: "Path to a specific .env file or directory to use as the entry point"
    }
  },
  examples: `
Executes a command in a child process, injecting your resolved and validated environment
variables from your .env files. Useful when a code-level integration is not possible.

Examples:
  varlock run -- node app.js                    # Run a Node.js application
  varlock run -- python script.py               # Run a Python script
  varlock run -- sh -c 'echo $MY_VAR'           # Use shell expansion for env vars
  varlock run --no-redact-stdout -- psql        # Preserve TTY for interactive tools
  varlock run --path .env.prod -- node app.js   # Use a specific .env file
  varlock run --path ./config/ -- node app.js   # Use a specific directory

\u{1F4CD} Important: Use -- to separate varlock options from your command

\u{1F4A1} Tip: For shell expansion of env vars, use: sh -c 'your command here'
\u{1F4A1} Tip: Use --no-redact-stdout for interactive tools (psql, claude, etc.)
  `.trim()
});
var commandProcess;
var childCommandKilledFromRestart = false;
var commandFn = /* @__PURE__ */ __name(async (ctx) => {
  const argv = process.argv.slice(2);
  let restCommandArgs = [];
  if (argv.includes("--")) {
    const doubleDashIndex = argv.indexOf("--");
    restCommandArgs = argv.slice(doubleDashIndex + 1);
  } else {
    throw new Error("No command to run! Your command should look like `varlock run -- <your-command>`");
  }
  const commandToRunAsArgs = restCommandArgs;
  const commandToRunStr = restCommandArgs.join(" ");
  const rawCommand = commandToRunAsArgs[0];
  const commandArgsOnly = commandToRunAsArgs.slice(1);
  const envGraph = await loadVarlockEnvGraph({
    entryFilePath: ctx.values.path
  });
  checkForSchemaErrors(envGraph);
  checkForNoEnvFiles(envGraph);
  await envGraph.generateTypesIfNeeded();
  await envGraph.resolveEnvValues();
  checkForConfigErrors(envGraph);
  const resolvedEnv = envGraph.getResolvedEnvObject();
  const serializedGraph = envGraph.getSerializedGraph();
  const fullInjectedEnv = {
    ...process.env,
    ...resolvedEnv,
    __VARLOCK_RUN: "1",
    // flag for a child process to detect it is runnign via `varlock run`
    __VARLOCK_ENV: JSON.stringify(serializedGraph)
  };
  const redactLogs = serializedGraph.settings?.redactLogs ?? true;
  const noRedactStdout = ctx.values["no-redact-stdout"] ?? false;
  if (redactLogs) {
    resetRedactionMap(serializedGraph);
  }
  if (noRedactStdout) {
    commandProcess = exec(rawCommand, commandArgsOnly, {
      stdio: "inherit",
      env: fullInjectedEnv
    });
  } else {
    const writeRedacted = /* @__PURE__ */ __name((stream, chunk) => {
      const str = chunk.toString();
      stream.write(redactLogs ? redactSensitiveConfig(str) : str);
    }, "writeRedacted");
    commandProcess = exec(rawCommand, commandArgsOnly, {
      stdin: "inherit",
      stdout: "pipe",
      stderr: "pipe",
      env: fullInjectedEnv
    });
    commandProcess.stdout?.on("data", (chunk) => writeRedacted(process.stdout, chunk));
    commandProcess.stderr?.on("data", (chunk) => writeRedacted(process.stderr, chunk));
  }
  {
    process.on("exit", (_code, _signal) => {
      commandProcess?.kill(9);
    });
    ["SIGTERM", "SIGINT"].forEach((signal) => {
      process.on(signal, () => {
        commandProcess?.kill(9);
        gracefulExit(1);
      });
    });
  }
  let exitCode;
  try {
    const result = await commandProcess;
    exitCode = result.exitCode;
  } catch (error) {
    if (error.signal === "SIGINT" && childCommandKilledFromRestart) ;
    if (error.signal === "SIGINT" || error.signal === "SIGKILL") {
      gracefulExit(1);
    } else {
      console.log(error.message);
      console.log(`command [${commandToRunStr}] failed`);
      console.log("try running the same command without varlock");
      console.log("if you get a different result, varlock may be the problem...");
    }
    exitCode = error.exitCode || 1;
  }
  {
    return gracefulExit(exitCode);
  }
}, "commandFn");

export { commandFn, commandSpec };
