// varlock/dist/chunk-Y3ITSQA4.js
import { __name } from './chunk-6PEHRAEP.js';
import { exec, spawn } from 'child_process';
import { promisify } from 'util';

var ExecError = class extends Error {
  constructor(exitCode, signal, data = "command gave no output") {
    super(data);
    this.exitCode = exitCode;
    this.signal = signal;
    this.data = data;
  }
  static {
    __name(this, "ExecError");
  }
};
function spawnAsyncHelper(command, args, spawnOptions) {
  const childProcess = spawn(command, args, spawnOptions || {});
  const deferred = new Promise((resolve, reject) => {
    let stdoutData = "";
    let stderrData = "";
    childProcess.stdout?.on("data", (data) => {
      stdoutData += data.toString();
    });
    childProcess.stderr?.on("data", (data) => {
      stderrData += data.toString();
    });
    childProcess.stdout?.on("error", (err) => {
      reject(err);
    });
    childProcess.stderr?.on("error", (err) => {
      reject(err);
    });
    childProcess.on("error", (err) => {
      reject(err);
    });
    childProcess.on("exit", (exitCode, signal) => {
      if (!exitCode) {
        resolve(stdoutData);
      } else {
        reject(
          new ExecError(exitCode, signal, stderrData)
        );
      }
    });
  });
  return { childProcess, execResult: deferred };
}
__name(spawnAsyncHelper, "spawnAsyncHelper");
async function spawnAsync(command, args, opts) {
  const { execResult } = spawnAsyncHelper(command, args, opts);
  return execResult;
}
__name(spawnAsync, "spawnAsync");
promisify(exec);

export { spawnAsync };
