// varlock/dist/chunk-HNG2WVGF.js
import { CliExitError } from './chunk-3D6G5PKR.js';
import { define } from './chunk-4A54P4EM.js';
import { checkForSchemaErrors } from './chunk-YBSEQARI.js';
import { loadVarlockEnvGraph } from './chunk-HSRXSFTK.js';
import { gracefulExit } from './chunk-PPIQGMSC.js';
import { __name } from './chunk-6PEHRAEP.js';

// src/cli/commands/printenv.command.ts
var commandSpec = define({
  name: "printenv",
  description: "Print the resolved value of a single environment variable",
  args: {
    path: {
      type: "string",
      short: "p",
      description: "Path to a specific .env file or directory (with trailing slash) to use as the entry point"
    }
  },
  examples: `
Prints the resolved value of a single environment variable.
Useful within larger shell commands where you need a single env var value.

Examples:
  varlock printenv MY_VAR                    # Print the value of MY_VAR
  varlock printenv --path .env.prod MY_VAR   # Use a specific .env file
  varlock printenv --path ./config/ MY_VAR   # Use a specific directory

\u{1F4CD} Note: Use sh -c to embed this in shell commands, e.g.:
       sh -c 'do-something --token $(varlock printenv MY_TOKEN)'

\u{1F4A1} Tip: Unlike \`varlock run -- echo $MY_VAR\`, this works because the shell
       expansion happens after varlock has printed the value.
  `.trim()
});
var commandFn = /* @__PURE__ */ __name(async (ctx) => {
  const positionals = (ctx.positionals ?? []).slice(ctx.commandPath?.length ?? 0);
  if (!positionals.length) {
    throw new CliExitError("Missing required argument: variable name", {
      suggestion: "Run `varlock printenv MY_VAR` to print the value of MY_VAR"
    });
  }
  const varName = positionals[0];
  const envGraph = await loadVarlockEnvGraph({
    entryFilePath: ctx.values.path
  });
  checkForSchemaErrors(envGraph);
  if (!(varName in envGraph.configSchema)) {
    throw new CliExitError(`Variable "${varName}" not found in schema`);
  }
  await envGraph.resolveItemWithDeps(varName);
  const item = envGraph.configSchema[varName];
  if (item.validationState === "error") {
    for (const err of item.errors) {
      console.error(`\u{1F6A8} ${err.message}`);
    }
    return gracefulExit(1);
  }
  const value = item.resolvedValue;
  if (value === void 0 || value === null) {
    console.log("");
  } else {
    console.log(String(value));
  }
}, "commandFn");

export { commandFn, commandSpec };
