// varlock/dist/chunk-PPIQGMSC.js
import { EnvGraph, DirectoryDataSource, DotEnvFileDataSource, my_dash_default, ansis_default } from './chunk-CBO5NXPJ.js';
import { redactString } from './chunk-XLYSNOR3.js';
import { __name } from './chunk-6PEHRAEP.js';
import process2 from 'process';
import fs from 'fs';
import path from 'path';

var asyncCallbacks = /* @__PURE__ */ new Set();
var callbacks = /* @__PURE__ */ new Set();
var isCalled = false;
var isRegistered = false;
async function flushStdio() {
  const flush = /* @__PURE__ */ __name((stream) => new Promise((resolve) => {
    if (!stream || !stream.writable || stream.writableEnded || stream.destroyed) {
      resolve();
      return;
    }
    const onError = /* @__PURE__ */ __name(() => {
      stream.off("error", onError);
      resolve();
    }, "onError");
    stream.once("error", onError);
    try {
      stream.write("", () => {
        stream.off("error", onError);
        resolve();
      });
    } catch {
      stream.off("error", onError);
      resolve();
    }
  }), "flush");
  const timeout = new Promise((resolve) => {
    setTimeout(resolve, 1e3);
  });
  await Promise.race([
    Promise.all([flush(process2.stdout), flush(process2.stderr)]),
    timeout
  ]);
}
__name(flushStdio, "flushStdio");
async function exit(shouldManuallyExit, isSynchronous, signal) {
  if (isCalled) {
    return;
  }
  isCalled = true;
  if (asyncCallbacks.size > 0 && isSynchronous) {
    console.error([
      "SYNCHRONOUS TERMINATION NOTICE:",
      "When explicitly exiting the process via process.exit or via a parent process,",
      "asynchronous tasks in your exitHooks will not run. Either remove these tasks,",
      "use gracefulExit() instead of process.exit(), or ensure your parent process",
      "sends a SIGINT to the process running this code."
    ].join(" "));
  }
  let exitCode = 0;
  if (signal > 0) {
    exitCode = 128 + signal;
  } else if (typeof process2.exitCode === "number" || typeof process2.exitCode === "string") {
    exitCode = process2.exitCode;
  }
  const done = /* @__PURE__ */ __name((force = false) => {
    if (force === true || shouldManuallyExit === true) {
      process2.exit(exitCode);
    }
  }, "done");
  for (const callback of callbacks) {
    callback(exitCode);
  }
  if (isSynchronous) {
    done();
    return;
  }
  const promises = [];
  let forceAfter = 0;
  for (const [callback, wait] of asyncCallbacks) {
    forceAfter = Math.max(forceAfter, wait);
    promises.push(Promise.resolve(callback(exitCode)));
  }
  const asyncTimer = forceAfter > 0 ? setTimeout(() => {
    done(true);
  }, forceAfter) : void 0;
  await Promise.all(promises);
  clearTimeout(asyncTimer);
  await flushStdio();
  done();
}
__name(exit, "exit");
function addHook(options) {
  const { onExit, wait, isSynchronous } = options;
  const asyncCallbackConfig = [onExit, wait];
  if (isSynchronous) {
    callbacks.add(onExit);
  } else {
    asyncCallbacks.add(asyncCallbackConfig);
  }
  if (!isRegistered) {
    isRegistered = true;
    process2.once("beforeExit", exit.bind(void 0, true, false, -128));
    process2.once("SIGINT", exit.bind(void 0, true, false, 2));
    process2.once("SIGTERM", exit.bind(void 0, true, false, 15));
    process2.once("exit", exit.bind(void 0, false, true, 0));
    process2.on("message", (message) => {
      if (message === "shutdown") {
        exit(true, true, -128);
      }
    });
  }
  return () => {
    if (isSynchronous) {
      callbacks.delete(onExit);
    } else {
      asyncCallbacks.delete(asyncCallbackConfig);
    }
  };
}
__name(addHook, "addHook");
function asyncExitHook(onExit, options = {}) {
  if (typeof onExit !== "function") {
    throw new TypeError("onExit must be a function");
  }
  if (!(typeof options.wait === "number" && options.wait > 0)) {
    throw new TypeError("wait must be set to a positive numeric value");
  }
  return addHook({
    onExit,
    wait: options.wait,
    isSynchronous: false
  });
}
__name(asyncExitHook, "asyncExitHook");
function gracefulExit(signal) {
  if (signal !== void 0) {
    process2.exitCode = signal;
  }
  exit(true, false, -128);
}
__name(gracefulExit, "gracefulExit");
async function loadEnvGraph(opts) {
  const graph = new EnvGraph();
  if (opts?.entryFilePath) {
    const resolvedPath = path.resolve(opts.entryFilePath);
    const isDirectory = opts.entryFilePath.endsWith("/") || opts.entryFilePath.endsWith(path.sep) || fs.existsSync(resolvedPath) && fs.statSync(resolvedPath).isDirectory();
    if (isDirectory) {
      graph.basePath = resolvedPath;
      if (opts?.afterInit) await opts.afterInit(graph);
      if (opts?.currentEnvFallback) graph.envFlagFallback = opts.currentEnvFallback;
      await graph.setRootDataSource(new DirectoryDataSource(resolvedPath));
    } else {
      graph.basePath = path.dirname(resolvedPath);
      if (opts?.afterInit) await opts.afterInit(graph);
      if (opts?.currentEnvFallback) graph.envFlagFallback = opts.currentEnvFallback;
      await graph.setRootDataSource(new DotEnvFileDataSource(resolvedPath));
    }
  } else {
    graph.basePath = opts?.basePath ?? process.cwd();
    if (opts?.afterInit) await opts.afterInit(graph);
    if (opts?.currentEnvFallback) graph.envFlagFallback = opts.currentEnvFallback;
    await graph.setRootDataSource(new DirectoryDataSource(graph.basePath));
  }
  await graph.finishLoad();
  return graph;
}
__name(loadEnvGraph, "loadEnvGraph");

// src/lib/formatting.ts
function applyMods(str, mods) {
  if (!mods) return str;
  if (my_dash_default.isArray(mods)) {
    let modStr = str;
    mods.forEach((mod) => {
      modStr = ansis_default[mod](modStr);
    });
    return modStr;
  }
  return ansis_default[mods](str);
}
__name(applyMods, "applyMods");
function formattedValue(val, showType = false) {
  let strVal = "";
  let strType = "";
  let mods;
  if (my_dash_default.isBoolean(val)) {
    strVal = val.toString();
    mods = ["yellow", "italic"];
    strType = "boolean";
  } else if (my_dash_default.isNumber(val)) {
    strVal = val.toString();
    mods = "yellow";
    strType = "number";
  } else if (my_dash_default.isString(val)) {
    strVal = `"${val}"`;
    strType = "string";
  } else if (my_dash_default.isPlainObject(val)) {
    strVal = JSON.stringify(val);
    strType = "object";
  } else if (val === null) {
    strVal = "null";
    mods = "gray";
  } else if (val === void 0) {
    strVal = "undefined";
    mods = "gray";
  }
  return [
    applyMods(strVal, mods),
    showType && strType ? ansis_default.gray(` (${strType})`) : ""
  ].join("");
}
__name(formattedValue, "formattedValue");
function joinAndCompact(strings, joinChar = " ") {
  return strings.filter((s) => (
    // we'll not filter out empty strings - because it's useful to just add newlines
    s !== void 0 && s !== null && s !== false
  )).join(joinChar);
}
__name(joinAndCompact, "joinAndCompact");
var VALIDATION_STATE_COLORS = {
  error: "red",
  warn: "yellow",
  valid: "cyan"
};
function getItemSummary(item) {
  const summary = [];
  const itemErrors = item.errors;
  const icon = itemErrors.length ? itemErrors[0].icon : "\u2705";
  const isSensitive = item.isSensitive;
  const isRequired = item.isRequired;
  summary.push(joinAndCompact([
    icon,
    ansis_default[VALIDATION_STATE_COLORS[item.validationState]](item.key) + (isRequired ? ansis_default.magenta("*") : ""),
    // ansis.gray(`[type = ${item.type.typeLabel}]`),
    isSensitive && ` \u{1F510}${ansis_default.gray.italic("sensitive")}`
    // item.useAt ? ansis.gray.italic(`(${item.useAt?.join(', ')})`) : undefined,
  ]));
  let valAsStr = formattedValue(item.resolvedValue, false);
  if (isSensitive && item.resolvedValue && my_dash_default.isString(item.resolvedValue)) {
    valAsStr = redactString(item.resolvedValue);
  }
  summary.push(joinAndCompact([
    ansis_default.gray("   \u2514"),
    valAsStr,
    item.isCoerced && ansis_default.gray.italic("< coerced from ") + (isSensitive ? formattedValue(item.resolvedRawValue) : formattedValue(item.resolvedRawValue, false))
  ]));
  itemErrors?.forEach((err) => {
    summary.push(ansis_default[err.isWarning ? "yellow" : "red"](`   - ${err.isWarning ? "[WARNING] " : ""}${err.message}`));
    if (err.tip) {
      summary.push(...err.tip.split("\n").map((line) => `     ${line}`));
    }
  });
  return summary.join("\n");
}
__name(getItemSummary, "getItemSummary");

export { asyncExitHook, getItemSummary, gracefulExit, joinAndCompact, loadEnvGraph };
