// varlock/dist/index.js
export { patchGlobalServerResponse } from './chunk-MYHVSJ3X.js';
import { checkBunVersion } from './chunk-ZIMSE3Z5.js';
import { checkForConfigErrors } from './chunk-YBSEQARI.js';
import { loadVarlockEnvGraph } from './chunk-HSRXSFTK.js';
import { loadEnvGraph } from './chunk-PPIQGMSC.js';
import { ResolutionError, CoercionError, ValidationError, SchemaError, ConfigLoadError, DotEnvFileDataSource, EnvGraph } from './chunk-CBO5NXPJ.js';
export { createDebug } from './chunk-QZ6HBRJC.js';
export { patchGlobalConsole } from './chunk-BC7LU4LG.js';
export { patchGlobalResponse } from './chunk-LVZSZAKN.js';
import { initVarlockEnv } from './chunk-IMB5QAZS.js';
export { ENV } from './chunk-IMB5QAZS.js';
import './chunk-XLYSNOR3.js';
import { __name } from './chunk-6PEHRAEP.js';

// src/index.ts
async function load() {
  checkBunVersion();
  const envGraph = await loadVarlockEnvGraph();
  await envGraph.resolveEnvValues();
  checkForConfigErrors(envGraph);
  process.env.__VARLOCK_ENV = JSON.stringify(envGraph.getSerializedGraph());
  initVarlockEnv();
}
__name(load, "load");
function getBuildTimeReplacements(opts) {
  if (!process.env.__VARLOCK_ENV) return {};
  const envInfo = JSON.parse(process.env.__VARLOCK_ENV);
  const replacements = {};
  for (const key in envInfo.config) {
    const itemInfo = envInfo.config[key];
    const replaceItem = !itemInfo.isSensitive || opts?.includeSensitive;
    if (!replaceItem) continue;
    replacements[`${opts?.objectKey || "ENV"}.${key}`] = JSON.stringify(envInfo.config[key].value);
  }
  return replacements;
}
__name(getBuildTimeReplacements, "getBuildTimeReplacements");
var internal = {
  // Core classes
  EnvGraph,
  DotEnvFileDataSource,
  // Loader function
  loadEnvGraph,
  // Error classes
  ConfigLoadError,
  SchemaError,
  ValidationError,
  CoercionError,
  ResolutionError,
  // Varlock-specific utilities
  loadVarlockEnvGraph,
  checkForConfigErrors,
  initVarlockEnv
};

export { getBuildTimeReplacements, internal, load };
