// varlock/dist/cli/cli-executable.js
import { commandSpec as commandSpec$5 } from '../chunk-VV7IRAYL.js';
import { commandSpec as commandSpec$6 } from '../chunk-AWPEMNMA.js';
import { commandSpec as commandSpec$7 } from '../chunk-J3H553VE.js';
import { commandSpec } from '../chunk-PGDFZB7I.js';
import '../chunk-Y3ITSQA4.js';
import { fmt } from '../chunk-343AECT3.js';
import { commandSpec as commandSpec$1 } from '../chunk-6KLPP3K7.js';
import { commandSpec as commandSpec$2 } from '../chunk-7B55HOP6.js';
import { commandSpec as commandSpec$3 } from '../chunk-X4Y5PWYI.js';
import { commandSpec as commandSpec$4 } from '../chunk-GJ7PTJM4.js';
import { cli } from '../chunk-4A54P4EM.js';
import { checkBunVersion } from '../chunk-BLY5NBV7.js';
import { InvalidEnvError } from '../chunk-COH4YYZ3.js';
import '../chunk-QCECXSMG.js';
import { CliExitError, gracefulExit, asyncExitHook } from '../chunk-ETMUQ3YH.js';
import { ansis_default, getUserVarlockDir } from '../chunk-MXPBXRWY.js';
import { createDebug } from '../chunk-QZ6HBRJC.js';
import '../chunk-IMB5QAZS.js';
import '../chunk-XLYSNOR3.js';
import { __commonJS, __name, __toESM } from '../chunk-6PEHRAEP.js';
import os2 from 'os';
import crypto, { createHash } from 'crypto';
import { join, dirname } from 'path';
import fs, { existsSync, readFileSync, mkdirSync, writeFileSync } from 'fs';
import process2 from 'process';

// ../../node_modules/.bun/ci-info@4.4.0/node_modules/ci-info/vendors.json
var require_vendors = __commonJS({
  "../../node_modules/.bun/ci-info@4.4.0/node_modules/ci-info/vendors.json"(exports$1, module) {
    module.exports = [
      {
        name: "Agola CI",
        constant: "AGOLA",
        env: "AGOLA_GIT_REF",
        pr: "AGOLA_PULL_REQUEST_ID"
      },
      {
        name: "Alpic",
        constant: "ALPIC",
        env: "ALPIC_HOST"
      },
      {
        name: "Appcircle",
        constant: "APPCIRCLE",
        env: "AC_APPCIRCLE",
        pr: {
          env: "AC_GIT_PR",
          ne: "false"
        }
      },
      {
        name: "AppVeyor",
        constant: "APPVEYOR",
        env: "APPVEYOR",
        pr: "APPVEYOR_PULL_REQUEST_NUMBER"
      },
      {
        name: "AWS CodeBuild",
        constant: "CODEBUILD",
        env: "CODEBUILD_BUILD_ARN",
        pr: {
          env: "CODEBUILD_WEBHOOK_EVENT",
          any: [
            "PULL_REQUEST_CREATED",
            "PULL_REQUEST_UPDATED",
            "PULL_REQUEST_REOPENED"
          ]
        }
      },
      {
        name: "Azure Pipelines",
        constant: "AZURE_PIPELINES",
        env: "TF_BUILD",
        pr: {
          BUILD_REASON: "PullRequest"
        }
      },
      {
        name: "Bamboo",
        constant: "BAMBOO",
        env: "bamboo_planKey"
      },
      {
        name: "Bitbucket Pipelines",
        constant: "BITBUCKET",
        env: "BITBUCKET_COMMIT",
        pr: "BITBUCKET_PR_ID"
      },
      {
        name: "Bitrise",
        constant: "BITRISE",
        env: "BITRISE_IO",
        pr: "BITRISE_PULL_REQUEST"
      },
      {
        name: "Buddy",
        constant: "BUDDY",
        env: "BUDDY_WORKSPACE_ID",
        pr: "BUDDY_EXECUTION_PULL_REQUEST_ID"
      },
      {
        name: "Buildkite",
        constant: "BUILDKITE",
        env: "BUILDKITE",
        pr: {
          env: "BUILDKITE_PULL_REQUEST",
          ne: "false"
        }
      },
      {
        name: "CircleCI",
        constant: "CIRCLE",
        env: "CIRCLECI",
        pr: "CIRCLE_PULL_REQUEST"
      },
      {
        name: "Cirrus CI",
        constant: "CIRRUS",
        env: "CIRRUS_CI",
        pr: "CIRRUS_PR"
      },
      {
        name: "Cloudflare Pages",
        constant: "CLOUDFLARE_PAGES",
        env: "CF_PAGES"
      },
      {
        name: "Cloudflare Workers",
        constant: "CLOUDFLARE_WORKERS",
        env: "WORKERS_CI"
      },
      {
        name: "Codefresh",
        constant: "CODEFRESH",
        env: "CF_BUILD_ID",
        pr: {
          any: [
            "CF_PULL_REQUEST_NUMBER",
            "CF_PULL_REQUEST_ID"
          ]
        }
      },
      {
        name: "Codemagic",
        constant: "CODEMAGIC",
        env: "CM_BUILD_ID",
        pr: "CM_PULL_REQUEST"
      },
      {
        name: "Codeship",
        constant: "CODESHIP",
        env: {
          CI_NAME: "codeship"
        }
      },
      {
        name: "Drone",
        constant: "DRONE",
        env: "DRONE",
        pr: {
          DRONE_BUILD_EVENT: "pull_request"
        }
      },
      {
        name: "dsari",
        constant: "DSARI",
        env: "DSARI"
      },
      {
        name: "Earthly",
        constant: "EARTHLY",
        env: "EARTHLY_CI"
      },
      {
        name: "Expo Application Services",
        constant: "EAS",
        env: "EAS_BUILD"
      },
      {
        name: "Gerrit",
        constant: "GERRIT",
        env: "GERRIT_PROJECT"
      },
      {
        name: "Gitea Actions",
        constant: "GITEA_ACTIONS",
        env: "GITEA_ACTIONS"
      },
      {
        name: "GitHub Actions",
        constant: "GITHUB_ACTIONS",
        env: "GITHUB_ACTIONS",
        pr: {
          GITHUB_EVENT_NAME: "pull_request"
        }
      },
      {
        name: "GitLab CI",
        constant: "GITLAB",
        env: "GITLAB_CI",
        pr: "CI_MERGE_REQUEST_ID"
      },
      {
        name: "GoCD",
        constant: "GOCD",
        env: "GO_PIPELINE_LABEL"
      },
      {
        name: "Google Cloud Build",
        constant: "GOOGLE_CLOUD_BUILD",
        env: "BUILDER_OUTPUT"
      },
      {
        name: "Harness CI",
        constant: "HARNESS",
        env: "HARNESS_BUILD_ID"
      },
      {
        name: "Heroku",
        constant: "HEROKU",
        env: {
          env: "NODE",
          includes: "/app/.heroku/node/bin/node"
        }
      },
      {
        name: "Hudson",
        constant: "HUDSON",
        env: "HUDSON_URL"
      },
      {
        name: "Jenkins",
        constant: "JENKINS",
        env: [
          "JENKINS_URL",
          "BUILD_ID"
        ],
        pr: {
          any: [
            "ghprbPullId",
            "CHANGE_ID"
          ]
        }
      },
      {
        name: "LayerCI",
        constant: "LAYERCI",
        env: "LAYERCI",
        pr: "LAYERCI_PULL_REQUEST"
      },
      {
        name: "Magnum CI",
        constant: "MAGNUM",
        env: "MAGNUM"
      },
      {
        name: "Netlify CI",
        constant: "NETLIFY",
        env: "NETLIFY",
        pr: {
          env: "PULL_REQUEST",
          ne: "false"
        }
      },
      {
        name: "Nevercode",
        constant: "NEVERCODE",
        env: "NEVERCODE",
        pr: {
          env: "NEVERCODE_PULL_REQUEST",
          ne: "false"
        }
      },
      {
        name: "Prow",
        constant: "PROW",
        env: "PROW_JOB_ID"
      },
      {
        name: "ReleaseHub",
        constant: "RELEASEHUB",
        env: "RELEASE_BUILD_ID"
      },
      {
        name: "Render",
        constant: "RENDER",
        env: "RENDER",
        pr: {
          IS_PULL_REQUEST: "true"
        }
      },
      {
        name: "Sail CI",
        constant: "SAIL",
        env: "SAILCI",
        pr: "SAIL_PULL_REQUEST_NUMBER"
      },
      {
        name: "Screwdriver",
        constant: "SCREWDRIVER",
        env: "SCREWDRIVER",
        pr: {
          env: "SD_PULL_REQUEST",
          ne: "false"
        }
      },
      {
        name: "Semaphore",
        constant: "SEMAPHORE",
        env: "SEMAPHORE",
        pr: "PULL_REQUEST_NUMBER"
      },
      {
        name: "Sourcehut",
        constant: "SOURCEHUT",
        env: {
          CI_NAME: "sourcehut"
        }
      },
      {
        name: "Strider CD",
        constant: "STRIDER",
        env: "STRIDER"
      },
      {
        name: "TaskCluster",
        constant: "TASKCLUSTER",
        env: [
          "TASK_ID",
          "RUN_ID"
        ]
      },
      {
        name: "TeamCity",
        constant: "TEAMCITY",
        env: "TEAMCITY_VERSION"
      },
      {
        name: "Travis CI",
        constant: "TRAVIS",
        env: "TRAVIS",
        pr: {
          env: "TRAVIS_PULL_REQUEST",
          ne: "false"
        }
      },
      {
        name: "Vela",
        constant: "VELA",
        env: "VELA",
        pr: {
          VELA_PULL_REQUEST: "1"
        }
      },
      {
        name: "Vercel",
        constant: "VERCEL",
        env: {
          any: [
            "NOW_BUILDER",
            "VERCEL"
          ]
        },
        pr: "VERCEL_GIT_PULL_REQUEST_ID"
      },
      {
        name: "Visual Studio App Center",
        constant: "APPCENTER",
        env: "APPCENTER_BUILD_ID"
      },
      {
        name: "Woodpecker",
        constant: "WOODPECKER",
        env: {
          CI: "woodpecker"
        },
        pr: {
          CI_BUILD_EVENT: "pull_request"
        }
      },
      {
        name: "Xcode Cloud",
        constant: "XCODE_CLOUD",
        env: "CI_XCODE_PROJECT",
        pr: "CI_PULL_REQUEST_NUMBER"
      },
      {
        name: "Xcode Server",
        constant: "XCODE_SERVER",
        env: "XCS"
      }
    ];
  }
});

// ../../node_modules/.bun/ci-info@4.4.0/node_modules/ci-info/index.js
var require_ci_info = __commonJS({
  "../../node_modules/.bun/ci-info@4.4.0/node_modules/ci-info/index.js"(exports$1) {
    var vendors = require_vendors();
    var env = process.env;
    Object.defineProperty(exports$1, "_vendors", {
      value: vendors.map(function(v) {
        return v.constant;
      })
    });
    exports$1.name = null;
    exports$1.isPR = null;
    exports$1.id = null;
    if (env.CI !== "false") {
      vendors.forEach(function(vendor) {
        const envs = Array.isArray(vendor.env) ? vendor.env : [vendor.env];
        const isCI2 = envs.every(function(obj) {
          return checkEnv(obj);
        });
        exports$1[vendor.constant] = isCI2;
        if (!isCI2) {
          return;
        }
        exports$1.name = vendor.name;
        exports$1.isPR = checkPR(vendor);
        exports$1.id = vendor.constant;
      });
    }
    exports$1.isCI = !!(env.CI !== "false" && // Bypass all checks if CI env is explicitly set to 'false'
    (env.BUILD_ID || // Jenkins, Cloudbees
    env.BUILD_NUMBER || // Jenkins, TeamCity
    env.CI || // Travis CI, CircleCI, Cirrus CI, Gitlab CI, Appveyor, CodeShip, dsari, Cloudflare Pages/Workers
    env.CI_APP_ID || // Appflow
    env.CI_BUILD_ID || // Appflow
    env.CI_BUILD_NUMBER || // Appflow
    env.CI_NAME || // Codeship and others
    env.CONTINUOUS_INTEGRATION || // Travis CI, Cirrus CI
    env.RUN_ID || // TaskCluster, dsari
    exports$1.name || false));
    function checkEnv(obj) {
      if (typeof obj === "string") return !!env[obj];
      if ("env" in obj) {
        return env[obj.env] && env[obj.env].includes(obj.includes);
      }
      if ("any" in obj) {
        return obj.any.some(function(k) {
          return !!env[k];
        });
      }
      return Object.keys(obj).every(function(k) {
        return env[k] === obj[k];
      });
    }
    __name(checkEnv, "checkEnv");
    function checkPR(vendor) {
      switch (typeof vendor.pr) {
        case "string":
          return !!env[vendor.pr];
        case "object":
          if ("env" in vendor.pr) {
            if ("any" in vendor.pr) {
              return vendor.pr.any.some(function(key) {
                return env[vendor.pr.env] === key;
              });
            } else {
              return vendor.pr.env in env && env[vendor.pr.env] !== vendor.pr.ne;
            }
          } else if ("any" in vendor.pr) {
            return vendor.pr.any.some(function(key) {
              return !!env[key];
            });
          } else {
            return checkEnv(vendor.pr);
          }
        default:
          return null;
      }
    }
    __name(checkPR, "checkPR");
  }
});
var VARLOCK_BANNER = [
  "",
  " \u2588\u2588\u2592   \u2588\u2593 \u2584\u2584\u2584       \u2588\u2588\u2580\u2588\u2588\u2588   \u2588\u2588\u2593     \u2592\u2588\u2588\u2588\u2588\u2588   \u2584\u2588\u2588\u2588\u2588\u2584   \u2588\u2588 \u2584\u2588\u2580",
  " \u2593\u2588\u2588\u2591   \u2588\u2592\u2592\u2588\u2588\u2588\u2588\u2584    \u2593\u2588\u2588 \u2592 \u2588\u2588\u2592\u2593\u2588\u2588\u2592    \u2592\u2588\u2588\u2592  \u2588\u2588\u2592\u2592\u2588\u2588\u2580 \u2580\u2588   \u2588\u2588\u2584\u2588\u2592 ",
  "  \u2593\u2588\u2588  \u2588\u2592\u2591\u2592\u2588\u2588  \u2580\u2588\u2584  \u2593\u2588\u2588 \u2591\u2584\u2588 \u2592\u2592\u2588\u2588\u2591    \u2592\u2588\u2588\u2591  \u2588\u2588\u2592\u2592\u2588\u2588    \u2584 \u2593\u2588\u2588\u2588\u2584\u2591 ",
  "   \u2592\u2588\u2588 \u2588\u2591\u2591\u2591\u2588\u2588\u2584\u2584\u2584\u2584\u2588\u2588 \u2592\u2588\u2588\u2580\u2580\u2588\u2584  \u2592\u2588\u2588\u2591    \u2592\u2588\u2588   \u2588\u2588\u2591\u2592\u2588\u2588\u2584 \u2584\u2588\u2588\u2592\u2593\u2588\u2588 \u2588\u2584 ",
  "    \u2592\u2580\u2588\u2591   \u2593\u2588   \u2593\u2588\u2588\u2592\u2591\u2588\u2588\u2593 \u2592\u2588\u2588\u2592\u2591\u2588\u2588\u2588\u2588\u2588\u2588\u2592\u2591 \u2588\u2588\u2588\u2588\u2588\u2592\u2591\u2592 \u2588\u2588\u2588\u2588\u2580 \u2591\u2592\u2588\u2588\u2592 \u2588\u2584",
  "    \u2591 \u2590\u2591   \u2592\u2592   \u2593\u2592\u2588\u2591\u2591 \u2592\u2593 \u2591\u2592\u2593\u2591\u2591 \u2592\u2591\u2593  \u2591\u2591 \u2592\u2591\u2592\u2591\u2592\u2591 \u2591 \u2591\u2592 \u2592  \u2591\u2592 \u2592\u2592 \u2593\u2592",
  "    \u2591 \u2591\u2591    \u2592   \u2592\u2592 \u2591  \u2591\u2592 \u2591 \u2592\u2591\u2591 \u2591 \u2592  \u2591  \u2591 \u2592 \u2592\u2591   \u2591  \u2592   \u2591 \u2591\u2592 \u2592\u2591",
  "      \u2591\u2591    \u2591   \u2592     \u2591\u2591   \u2591   \u2591 \u2591   \u2591 \u2591 \u2591 \u2592  \u2591        \u2591 \u2591\u2591 \u2591 ",
  "       \u2591        \u2591  \u2591   \u2591         \u2591  \u2591    \u2591 \u2591  \u2591 \u2591      \u2591  \u2591   "
  // '      ░                                       ░               ',
].join("\n");
var VARLOCK_BANNER_COLOR = VARLOCK_BANNER.replaceAll(/([▀▁▂▃▄▅▆▇█▉▊▋▌▍▎▏▐▔▕▖▗▘▙▚▛▜▝▞▟]+)/g, (m) => ansis_default.hex("#DD0000")(m)).replaceAll(/([▓░▒]+)/g, (m) => ansis_default.blue(m));
var import_ci_info = __toESM(require_ci_info(), 1);
var isDockerCached;
function hasDockerEnv() {
  try {
    fs.statSync("/.dockerenv");
    return true;
  } catch {
    return false;
  }
}
__name(hasDockerEnv, "hasDockerEnv");
function hasDockerCGroup() {
  try {
    return fs.readFileSync("/proc/self/cgroup", "utf8").includes("docker");
  } catch {
    return false;
  }
}
__name(hasDockerCGroup, "hasDockerCGroup");
function hasDockerMountInfo() {
  try {
    return fs.readFileSync("/proc/self/mountinfo", "utf8").includes("/docker/containers/");
  } catch {
    return false;
  }
}
__name(hasDockerMountInfo, "hasDockerMountInfo");
function isDocker() {
  isDockerCached ??= hasDockerEnv() || hasDockerCGroup() || hasDockerMountInfo();
  return isDockerCached;
}
__name(isDocker, "isDocker");
var isDockerCached2;
function hasDockerEnv2() {
  try {
    fs.statSync("/.dockerenv");
    return true;
  } catch {
    return false;
  }
}
__name(hasDockerEnv2, "hasDockerEnv");
function hasDockerCGroup2() {
  try {
    return fs.readFileSync("/proc/self/cgroup", "utf8").includes("docker");
  } catch {
    return false;
  }
}
__name(hasDockerCGroup2, "hasDockerCGroup");
function isDocker2() {
  if (isDockerCached2 === void 0) {
    isDockerCached2 = hasDockerEnv2() || hasDockerCGroup2();
  }
  return isDockerCached2;
}
__name(isDocker2, "isDocker");

// ../../node_modules/.bun/is-inside-container@1.0.0/node_modules/is-inside-container/index.js
var cachedResult;
var hasContainerEnv = /* @__PURE__ */ __name(() => {
  try {
    fs.statSync("/run/.containerenv");
    return true;
  } catch {
    return false;
  }
}, "hasContainerEnv");
function isInsideContainer() {
  if (cachedResult === void 0) {
    cachedResult = hasContainerEnv() || isDocker2();
  }
  return cachedResult;
}
__name(isInsideContainer, "isInsideContainer");

// ../../node_modules/.bun/is-wsl@3.1.1/node_modules/is-wsl/index.js
var isWsl = /* @__PURE__ */ __name(() => {
  if (process2.platform !== "linux") {
    return false;
  }
  if (os2.release().toLowerCase().includes("microsoft")) {
    if (isInsideContainer()) {
      return false;
    }
    return true;
  }
  try {
    if (fs.readFileSync("/proc/version", "utf8").toLowerCase().includes("microsoft")) {
      return !isInsideContainer();
    }
  } catch {
  }
  if (fs.existsSync("/proc/sys/fs/binfmt_misc/WSLInterop") || fs.existsSync("/run/WSL")) {
    return !isInsideContainer();
  }
  return false;
}, "isWsl");
var is_wsl_default = process2.env.__IS_WSL_TEST__ ? isWsl : isWsl();

// package.json
var package_default = {
  version: "0.6.4"};

// src/config.ts
var CONFIG = {
  // varlock.dev app
  POSTHOG_API_KEY: "phc_bfzH97VIta8yQa8HrsgmitqS6rTydjMISs0m8aqJTnq",
  POSTHOG_HOST: "https://ph.varlock.dev"
};

// src/cli/helpers/telemetry.ts
var debug = createDebug("varlock:telemetry");
var TRUE_ENV_VAR_VALUES = ["true", "1", "t"];
var userVarlockDirPath = getUserVarlockDir();
var userVarlockConfigFilePath = join(userVarlockDirPath, "config.json");
var userVarlockConfig;
var projectVarlockConfig;
var mergedVarlockConfigFileContents;
var _gitDirPath;
var _varlockDirPath;
var _projectRootDirPath;
var _foundProjectRoot = false;
function findProjectDirs() {
  if (!_foundProjectRoot) {
    let currentDir = process.cwd();
    while (currentDir) {
      const possibleGitDirPath = join(currentDir, ".git");
      if (!_gitDirPath && existsSync(possibleGitDirPath)) {
        _gitDirPath = possibleGitDirPath;
      }
      const possibleVarlockDirPath = join(currentDir, ".varlock");
      if (!_varlockDirPath && possibleVarlockDirPath !== userVarlockDirPath && existsSync(possibleVarlockDirPath)) {
        _varlockDirPath = possibleVarlockDirPath;
      }
      const parentDir = dirname(currentDir);
      if (parentDir === currentDir) break;
      currentDir = parentDir;
    }
    if (_gitDirPath) _projectRootDirPath = dirname(_gitDirPath);
    else if (_varlockDirPath) _projectRootDirPath = dirname(_varlockDirPath);
    else _projectRootDirPath = process.cwd();
    _foundProjectRoot = true;
  }
  return {
    gitDirPath: _gitDirPath,
    varlockDirPath: _varlockDirPath,
    projectRootDirPath: _projectRootDirPath
  };
}
__name(findProjectDirs, "findProjectDirs");
function loadVarlockConfig() {
  if (mergedVarlockConfigFileContents) return mergedVarlockConfigFileContents;
  try {
    const userConfigStr = readFileSync(userVarlockConfigFilePath, "utf-8");
    userVarlockConfig = userConfigStr.trim() ? JSON.parse(userConfigStr) : void 0;
  } catch (err) {
    if (err instanceof Error && "code" in err && err.code === "ENOENT") {
      debug(`User varlock config file not found - ${userVarlockConfigFilePath}`);
    } else if (err instanceof SyntaxError) {
      throw new Error(`Invalid JSON in project varlock config - ${userVarlockConfigFilePath}`, { cause: err });
    } else {
      throw new Error(`Problem reading project varlock config - ${userVarlockConfigFilePath}`, { cause: err });
    }
  }
  const { varlockDirPath } = findProjectDirs();
  if (varlockDirPath) {
    const projectVarlockConfigPath = join(varlockDirPath, "config.json");
    try {
      const projectConfigStr = readFileSync(projectVarlockConfigPath, "utf-8");
      projectVarlockConfig = projectConfigStr.trim() ? JSON.parse(projectConfigStr) : void 0;
    } catch (err) {
      if (err instanceof Error && "code" in err && err.code === "ENOENT") {
        debug(`Project varlock config file not found - ${projectVarlockConfigPath}`);
      } else if (err instanceof SyntaxError) {
        throw new Error(`Invalid JSON in project varlock config - ${projectVarlockConfigPath}`, { cause: err });
      } else {
        throw new Error(`Problem reading project varlock config - ${projectVarlockConfigPath}`, { cause: err });
      }
    }
  }
  mergedVarlockConfigFileContents = {
    ...userVarlockConfig,
    ...projectVarlockConfig
  };
  return mergedVarlockConfigFileContents;
}
__name(loadVarlockConfig, "loadVarlockConfig");
var cachedAnonymousId;
function getAnonymousId() {
  if (cachedAnonymousId) return cachedAnonymousId;
  const varlockConfig = loadVarlockConfig();
  if (varlockConfig?.anonymousId) {
    cachedAnonymousId = varlockConfig.anonymousId;
    return varlockConfig.anonymousId;
  }
  const newAnonymousId = `${import_ci_info.isCI ? "ci-" : ""}${crypto.randomUUID()}`;
  try {
    if (!existsSync(userVarlockDirPath)) {
      mkdirSync(userVarlockDirPath, { recursive: true });
    }
    writeFileSync(
      userVarlockConfigFilePath,
      JSON.stringify({
        ...userVarlockConfig,
        anonymousId: newAnonymousId
      }, null, 2),
      { flag: "w" }
    );
  } catch (err) {
    if (os2.homedir() === "/dev/null") {
      debug("HOME directory is not set (probably running in Docker without HOME configured). Cannot persist anonymous ID.");
    } else {
      debug(`Cannot write to varlock config folder (${userVarlockDirPath}): ${err.message}`);
    }
  }
  cachedAnonymousId = newAnonymousId;
  return newAnonymousId;
}
__name(getAnonymousId, "getAnonymousId");
function checkIsOptedOut() {
  if (process.env.PH_OPT_OUT === "true" || process.env.VARLOCK_TELEMETRY_DISABLED && TRUE_ENV_VAR_VALUES.includes(process.env.VARLOCK_TELEMETRY_DISABLED.toLowerCase())) {
    debug("telemetry opted out - env var");
    return true;
  }
  const varlockConfig = loadVarlockConfig();
  if (varlockConfig?.analytics_opt_out || varlockConfig?.telemetryDisabled) {
    debug(`telemetry opted out - config file (${projectVarlockConfig?.telemetryDisabled ? "project" : "user"} config)`);
    return true;
  }
  return false;
}
__name(checkIsOptedOut, "checkIsOptedOut");
function anonymizeValue(payload) {
  if (payload === "") {
    return payload;
  }
  const hash = createHash("sha256");
  hash.update(payload);
  return hash.digest("hex");
}
__name(anonymizeValue, "anonymizeValue");
function getProjectGitRemoteUrl() {
  findProjectDirs();
  if (!_gitDirPath) return void 0;
  try {
    const gitConfigContents = readFileSync(join(_gitDirPath, "config"), "utf-8");
    const remoteUpstreamPos = gitConfigContents.indexOf('[remote "upstream"]');
    if (remoteUpstreamPos !== -1) {
      const remoteUpstreamUrl = gitConfigContents.slice(remoteUpstreamPos).match(/url = (.+)/)?.[1];
      return remoteUpstreamUrl;
    }
    const remoteOriginPos = gitConfigContents.indexOf('[remote "origin"]');
    if (remoteOriginPos === -1) return void 0;
    const remoteOriginUrl = gitConfigContents.slice(remoteOriginPos).match(/url = (.+)/)?.[1];
    return remoteOriginUrl;
  } catch (err) {
    return void 0;
  }
}
__name(getProjectGitRemoteUrl, "getProjectGitRemoteUrl");
function getAnonymousProjectId() {
  const gitRemoteUrl = getProjectGitRemoteUrl();
  if (!gitRemoteUrl) return null;
  return anonymizeValue(gitRemoteUrl);
}
__name(getAnonymousProjectId, "getAnonymousProjectId");
var cachedTelemetryMetadata;
function getTelemetryMeta() {
  if (cachedTelemetryMetadata) return cachedTelemetryMetadata;
  const cpus = os2.cpus() || [];
  let versionIdentifier = package_default.version;
  cachedTelemetryMetadata = {
    anonymous_project_id: getAnonymousProjectId(),
    node_version: process.version.replace(/^v?/, ""),
    varlock_version: versionIdentifier,
    // TODO: pass through version info for specific integrations/plugins?
    system_platform: os2.platform(),
    system_release: os2.release(),
    system_architecture: os2.arch(),
    cpu_count: cpus.length,
    cpu_model: cpus.length ? cpus[0].model : null,
    cpu_speed: cpus.length ? cpus[0].speed : null,
    memory_mb: Math.trunc(os2.totalmem() / 1024 ** 2),
    is_docker: isDocker(),
    is_tty: process.stdout.isTTY,
    is_wsl: is_wsl_default,
    is_ci: import_ci_info.isCI,
    ci_name: import_ci_info.name,
    is_sea: false
  };
  return cachedTelemetryMetadata;
}
__name(getTelemetryMeta, "getTelemetryMeta");
var isOptedOut = checkIsOptedOut();
var lastTelemetryReq;
async function posthogCapture(event, properties) {
  const telemetryMeta = getTelemetryMeta();
  const payload = {
    api_key: CONFIG.POSTHOG_API_KEY,
    event,
    properties: {
      $process_person_profile: false,
      ...telemetryMeta,
      ...properties
    },
    distinct_id: isOptedOut ? "---" : getAnonymousId()
  };
  debug(`track${isOptedOut ? " (disabled)" : ""}`, payload);
  if (isOptedOut) return;
  const removeExitHook = asyncExitHook(async () => {
    await lastTelemetryReq;
  }, { wait: 500 });
  lastTelemetryReq = fetch(`${CONFIG.POSTHOG_HOST}/i/v0/e/`, {
    method: "POST",
    body: JSON.stringify(payload),
    headers: {
      "Content-Type": "application/json"
    }
  }).then((res) => {
    if (!res.ok) throw new Error(`HTTP error! status: ${res.status}`);
    return res.text();
  }).then((text) => debug("telemetry response:", text)).catch((error) => {
    debug("telemetry error:", error);
  }).finally(() => {
    removeExitHook();
  });
}
__name(posthogCapture, "posthogCapture");
async function trackCommand(command, properties) {
  await posthogCapture("cli_command_executed", {
    command,
    ...properties
  });
}
__name(trackCommand, "trackCommand");

// src/cli/cli-executable.ts
var versionId = package_default.version;
function buildLazyCommand(commandSpec9, loadCommandFn) {
  const commandName = commandSpec9.name;
  return {
    ...commandSpec9,
    run: /* @__PURE__ */ __name(async (...args) => {
      await trackCommand(commandName, { command: commandName });
      const commandSpecAndFn = await loadCommandFn();
      return commandSpecAndFn.commandFn(...args);
    }, "run")
  };
}
__name(buildLazyCommand, "buildLazyCommand");
var subCommands = /* @__PURE__ */ new Map();
subCommands.set("init", buildLazyCommand(commandSpec, async () => await import('../init.command-IS5TWLU4.js')));
subCommands.set("load", buildLazyCommand(commandSpec$1, async () => await import('../load.command-OXPEKYKF.js')));
subCommands.set("run", buildLazyCommand(commandSpec$2, async () => await import('../run.command-INBQRQIZ.js')));
subCommands.set("printenv", buildLazyCommand(commandSpec$3, async () => await import('../printenv.command-DBEAAYGI.js')));
subCommands.set("help", buildLazyCommand(commandSpec$4, async () => await import('../help.command-7E52XAOO.js')));
subCommands.set("telemetry", buildLazyCommand(commandSpec$5, async () => await import('../telemetry.command-EYJSX3J7.js')));
subCommands.set("scan", buildLazyCommand(commandSpec$6, async () => await import('../scan.command-AE6OUPX2.js')));
subCommands.set("typegen", buildLazyCommand(commandSpec$7, async () => await import('../typegen.command-H7PTAPGH.js')));
(/* @__PURE__ */ __name(async function go() {
  try {
    try {
      checkBunVersion();
    } catch (e) {
      throw new CliExitError(e.message, { forceExit: true });
    }
    let args = process.argv.slice(2);
    if (args[0] === "help") args = ["--help"];
    if (false) ;
    if (args[0] === "--version") {
      await trackCommand("version");
    }
    await cli(args, {
      // main command - triggered if you just run `varlock` with no args
      run: /* @__PURE__ */ __name(() => {
        console.log("Please run one of the sub-commands. Run `varlock --help` for more info.");
      }, "run")
    }, {
      name: "varlock",
      description: "Encrypt and protect your env vars",
      version: versionId,
      subCommands,
      renderHeader: /* @__PURE__ */ __name(async (ctx) => {
        if (ctx.name) return "";
        return VARLOCK_BANNER_COLOR;
      }, "renderHeader")
    });
    gracefulExit();
  } catch (error) {
    if (error instanceof Error && error.message.startsWith("Command not found: ")) {
      const badCommandName = error.message.split(": ")[1];
      const badCommandErr = new CliExitError(`Invalid subcommand: ${badCommandName}`, {
        suggestion: `Run \`${fmt.command("varlock --help", { jsPackageManager: true })}\` for more info.`
      });
      console.error(badCommandErr.getFormattedOutput());
    } else if (error instanceof CliExitError || error instanceof InvalidEnvError) {
      console.error(error.getFormattedOutput());
    } else {
      throw error;
    }
    gracefulExit(1);
  }
}, "go"))();
//! this ouput is used by homebrew formula to check installed version is correct
